mmv1/templates/terraform/post_update/tagtemplate_fields.go.erb

```diff
@@ -0,0 +1,114 @@
+// maps of old and new values
+for _, raw := range o.(*schema.Set).List() {
+    obj := raw.(map[string]interface{})
+    k := obj["field_id"].(string)
+    vals[k] = &FieldChange{Old: obj}
+}
+
+for _, raw := range n.(*schema.Set).List() {
+    obj := raw.(map[string]interface{})
+    k := obj["field_id"].(string)
+    if _, ok := vals[k]; !ok {
+        // if key is not present in the vals,
+        // then create an empty object to hold the new value
+        vals[k] = &FieldChange{}
+    }
+    vals[k].New = obj
+}
+
+// fields schema to create schema.set below
+dataCatalogTagTemplateFieldsSchema := &schema.Resource{
+    Schema: resourceDataCatalogTagTemplate().Schema["fields"].Elem.(*schema.Resource).Schema,
+}
+
+for name, change := range vals {
+    // A few different situations to deal with in here:
+    // - change.Old is nil: create a new role
+    // - change.New is nil: remove an existing role
+    // - both are set: test if New is different than Old and update if so
+
+    changeOldSet := schema.NewSet(schema.HashResource(dataCatalogTagTemplateFieldsSchema), []interface{}{})
+    changeOldSet.Add(change.Old)
+    var changeOldProp map[string]interface{}
+    if len(change.Old) != 0 {
+        changeOldProp, _ = expandDataCatalogTagTemplateFields(changeOldSet, nil, nil)
+        changeOldProp = changeOldProp[name].(map[string]interface{})
+    }
+
+    changeNewSet := schema.NewSet(schema.HashResource(dataCatalogTagTemplateFieldsSchema), []interface{}{})
+    changeNewSet.Add(change.New)
+    var changeNewProp map[string]interface{}
+    if len(change.New) != 0 {
+        changeNewProp, _ = expandDataCatalogTagTemplateFields(changeNewSet, nil, nil)
+        changeNewProp = changeNewProp[name].(map[string]interface{})
+    }
+
+    // if old state is empty, then we have a new field to create
+    if len(change.Old) == 0 {
+        err := createTagTemplateField(d, config, changeNewProp, name, billingProject, userAgent)
+        if err != nil {
+            return err
+        }
+
+        continue
+    }
+
+    // if new state is empty, then we need to delete the current field
+    if len(change.New) == 0 {
+        err := deleteTagTemplateField(d, config, name, billingProject, userAgent)
+        if err != nil {
+            return err
+        }
+
+        continue
+    }
+
+    // if we have old and new values, but are not equal, update with the new state
+    if !reflect.DeepEqual(changeOldProp, changeNewProp) {
+        url1, err := replaceVars(d, config, "{{DataCatalogBasePath}}{{name}}/fields/"+name)
+        if err != nil {
+            return err
+        }
+
+        oldType := changeOldProp["type"].(map[string]interface{})
+        newType := changeNewProp["type"].(map[string]interface{})
+
+        if oldType["primitiveType"] != newType["primitiveType"] {
+            // As primitiveType can't be change, consider it as ForceNew which triggers delete old field and recreate new field
```

Comment:
Would you mind adding a test on this? I tested it locally and got it worked in some cases but failed in others.
For example, if I try to change type of field `source` from `STRING` to `DOUBLE`, it ended in error `Error: Error creating TagTemplate Field: googleapi: Error 400: Adding a new required field on an existing tag template is not supported, for field source.`, but for field `num_rows` from `DOUBLE` to `STRING`, it worked fine. I wonder if it may have something to do with `is_required = true` being set?
The overall update + delete/recreate logic looks good to me though.

Reply:
Actually, I've tried the same use-case where it updates the type, it was working for me. Also, I've noticed the error message you put which says that it won't allow you to add any required field in the template once the template was created.

Reply:
I've added the [check](https://github.com/GoogleCloudPlatform/magic-modules/pull/6803/files#diff-19b28c738bebcea45ddae901685cf9cbb003c65e2f74a6f097b5f9eeae575860R91) for the required field while updating its primitiveType like ForceNew because if it is required then it will delete the field but it won't be able to create in previous code which led to inconsistent state. Hence, it is preferable to add a prior check for it, however if the user wants to update the primitiveType along-with making it optional, then it should work. Thanks for highlighting this!!

I've added the TCs for these as well.